api/V1/Recording.js

```diff
@@ -0,0 +1,201 @@
+var util = require('./util');
+var passport = require('passport');
+var log = require('../../logging');
+var tagsUtil = require('./tagsUtil');
+var requestUtil = require('./requestUtil');
+var responseUtil = require('./responseUtil');
+var multiparty = require('multiparty');
+var config = require('../../config/config')
+var randomstring = require("randomstring");
+var AWS = require('aws-sdk');
+
+module.exports = (app, baseUrl) => {
+  var apiUrl = baseUrl + '/recordings';
+
+  /**
+   * @api {post} /api/v1/recordings Add a new recording.
+   * @apiName PostRecording
+   * @apiGroup Recordings
+   * @apiDescription This call is the new way to upload any type of recording.
+   * It takes a 'data' field which contains a JSON object string that can
+   * contain any of the following fields, Note that 'type' is required:
+   * - type: REQUIRED, one of ('thermalRaw160x120')
+   * - duration
+   * - recordingDateTime
+   * - location
+   * - version
+   * - batteryCharging
+   * - batteryLevel
+   * - airplaneModeOn
+   * - additionalMetadata
+   * - processingMeta
+   *
+   * @apiUse V1DeviceAuthorizationHeader
+   *
+   * @apiParam {JSON} data Metadata about the recording (see above).
+   * @apiParam {File} file File of the recording.
+   *
+   * @apiUse V1ResponseSuccess
+   * @apiuse V1ResponseError
+   */
+  app.post(
+    apiUrl,
+    passport.authenticate(['jwt'], { session: false }),
+    async (request, response) => {
+      log.info(request.method + " Request: " + request.url);
+
+      // Check that the request was authenticated by a device.
+      if (request.user !== null && !requestUtil.isFromADevice(request))
+        return responseUtil.notFromADevice(response);
+      var device = request.user;
+
+      var recording;
+      var key = randomstring.generate();
+      var form = new multiparty.Form();
+      var validData = false;
+      var validFile = false;
+      var invalidMessages = [];
+
+      // Make new Recording from the data field and Device.
+      // TODO Stop stream if 'data' is invalid.
+      form.on('field', (name, value) => {
+        if (name != 'data') return; // Only parse data field.
+        try {
+          var data = JSON.parse(value);
+          recording = models.Recording.build(data, {
+            fields: models.Recording.apiSettableFields,
+          });
+          recording.set('rawFileKey', key);
+          recording.set('DeviceId', device.id);
+          recording.set('GroupId', device.GroupId);
+          recording.set('processingState',
+            models.Recording.processingStates[data.type][0]);
+          if (typeof device.public === 'boolean')
+            recording.set('public', device.public);
+          validData = true;
+        } catch (e) {
+          log.debug(e);
+          // TODO Stop stream here.
+        }
+      });
+
+      // Stream file to LeoFS.
+      form.on('part', (part) => {
+        if (part.name != 'file') return part.resume();
+        validFile = true;
```

Comment:
Should `validFile` only be set to `true` once `s3.upload` has completed successfully?